src/sparse_int_set.jl

```diff
@@ -0,0 +1,265 @@
+        s.counters[from_page] -= 1
+        pop!(s.packed)
+    end
+    cleanup!(s, from_page)
+    return id
+end
+
+function cleanup!(s::SparseIntSet, pageid::Int)
+    if s.counters[pageid] == 0
+        s.reverse[pageid] = NULL_INT_PAGE
+    end
+end
+
+function pop!(s::SparseIntSet, id::Integer, default)
+    id < 0 && throw(ArgumentError("Int to pop needs to be positive."))
+    return in(id, s) ? (@inbounds pop!(s, id)) : default
+end
+popfirst!(s::SparseIntSet) = pop!(s, first(s))
+
+iterate(set::SparseIntSet, args...) = iterate(set.packed, args...) 
+
+last(s::SparseIntSet) = isempty(s) ? throw(ArgumentError("Empty set has no last element.")) : last(s.packed)
+
+union(s::SparseIntSet, ns) = union!(copy(s), ns)
+function union!(s::SparseIntSet, ns)
+    for n in ns
+        push!(s, n)
+    end
+    return s
+end
+
+intersect(s1::SparseIntSet) = copy(s1)
+intersect(s1::SparseIntSet, ss...) = intersect(s1, intersect(ss...))
+function intersect(s1::SparseIntSet, ns)
+    s = SparseIntSet()
+    for n in ns
+        n in s1 && push!(s, n)
+    end
+    return s
+end
+
+intersect!(s1::SparseIntSet, ss...) = intersect!(s1, intersect(ss...))
+
+#Is there a more performant way to do this?
+intersect!(s1::SparseIntSet, ns) = copy!(s1, intersect(s1, ns))
+
+setdiff(s::SparseIntSet, ns) = setdiff!(copy(s), ns)
+function setdiff!(s::SparseIntSet, ns)
+    for n in ns
+        pop!(s, n, nothing)
+    end
+    return s
+end
+
+function ==(s1::SparseIntSet, s2::SparseIntSet)
+    length(s1) != length(s2) && return false
+    return all(in(s1), s2)
+end
+
+issubset(a::SparseIntSet, b::SparseIntSet) = isequal(a, intersect(a, b))
+
+complement(a::SparseIntSet) = complement!(SparseIntSet(), a)
+function complement!(b::SparseIntSet, a::SparseIntSet)
+    empty!(b)
+    for i in eachindex(a.reverse)
+        if a.reverse[i] === NULL_INT_PAGE
+            new_ids = ((i - 1) * INT_PER_PAGE + 1) : (i * INT_PER_PAGE)
+            append!(b.packed, new_ids)
+            if i > length(b.reverse)
+                push!(b.reverse, collect(new_ids)) 
+                push!(b.counters, INT_PER_PAGE)
+            else
+                b.reverse[i] = collect(new_ids)
+                b.counters[i] = INT_PER_PAGE
+            end
+        else
+            for offset in 1:INT_PER_PAGE
+                if a.reverse[i][offset] == 0
+                    push!(b, INT_PER_PAGE*(i-1) + offset)
+                end
+            end
+        end
+    end
+    return b
+end
+#Can this be optimized?
+complement!(a::SparseIntSet) = copy!(a, complement(a))
+
+<(a::SparseIntSet, b::SparseIntSet) = ( a<=b ) && !isequal(a, b)
+<=(a::SparseIntSet, b::SparseIntSet) = issubset(a, b)
+
+function findfirst_packed_id(i, s::SparseIntSet)
+    pageid, offset = pageid_offset(s, i)
+    if pageid > length(s.counters) || s.counters[pageid] == 0
+        return 0
+    end
+    @inbounds id = s.reverse[pageid][offset]
+    return id
+end
+
+collect(s::SparseIntSet) = s.packed
```

Comment:
Don't want to be returning something that could mutate the underlying data.

```suggestion
collect(s::SparseIntSet) = copy(s.packed)
```